vscode-extensions/playwright-on-codespaces-vscode-extension/src/extension.ts

```diff
@@ -0,0 +1,284 @@
+    const currentTunnel: PlaywrightTunnel | undefined = tunnel;
+    if (!currentTunnel) {
+      outputChannel.appendLine('No tunnel instance to stop.');
+      void vscode.window.showInformationMessage('Playwright tunnel is not running.');
+      return;
+    }
+
+    // Clear the reference before awaiting to avoid race condition
+    tunnel = undefined;
+
+    try {
+      outputChannel.appendLine('Stopping Playwright tunnel...');
+      await currentTunnel.stopAsync();
+      updateStatusBar('stopped');
+      outputChannel.appendLine('Tunnel stopped.');
+      void vscode.window.showInformationMessage('Playwright tunnel stopped.');
+    } catch (error) {
+      const errorMessage: string = error instanceof Error ? error.message : String(error);
+      outputChannel.appendLine(`Failed to stop tunnel: ${errorMessage}`);
+      void vscode.window.showErrorMessage(`Failed to stop Playwright tunnel: ${errorMessage}`);
+    }
+  }
+
+  async function handleShowMenu(): Promise<void> {
+    interface IQuickPickItem extends vscode.QuickPickItem {
+      action: 'start' | 'stop' | 'showLog';
+    }
+
+    const items: IQuickPickItem[] = [
+      {
+        label: '$(play) Start Tunnel',
+        description: 'Start the Playwright browser tunnel',
+        action: 'start'
+      },
+      {
+        label: '$(debug-stop) Stop Tunnel',
+        description: 'Stop the Playwright browser tunnel',
+        action: 'stop'
+      },
+      {
+        label: '$(output) Show Logs',
+        description: 'Show the Playwright tunnel output log',
+        action: 'showLog'
+      }
+    ];
+
+    const selected: IQuickPickItem | undefined = await vscode.window.showQuickPick(items, {
+      placeHolder: `Playwright Tunnel (${currentStatus})`
+    });
+
+    if (selected) {
+      switch (selected.action) {
+        case 'start':
+          await handleStartTunnel();
+          break;
+        case 'stop':
+          await handleStopTunnel();
+          break;
+        case 'showLog':
+          handleShowLog();
+          break;
+      }
+    }
+  }
+
+  context.subscriptions.push(
+    outputChannel,
+    statusBarItem,
+    vscode.commands.registerCommand(COMMAND_SHOW_LOG, handleShowLog),
+    vscode.commands.registerCommand(COMMAND_SHOW_SETTINGS, handleShowSettings),
+    vscode.commands.registerCommand(COMMAND_START_TUNNEL, handleStartTunnel),
+    vscode.commands.registerCommand(COMMAND_STOP_TUNNEL, handleStopTunnel),
+    vscode.commands.registerCommand(COMMAND_SHOW_MENU, handleShowMenu),
+    // Cleanup tunnel on deactivate
+    {
+      dispose: () => {
+        const currentTunnel: PlaywrightTunnel | undefined = tunnel;
+        if (currentTunnel) {
+          outputChannel.appendLine('Extension deactivating, stopping tunnel...');
+          void currentTunnel.stopAsync().then(() => {
+            tunnel = undefined;
+          });
+        }
+      }
+    }
+  );
+
+  // Auto-start the tunnel on activation
+  void handleStartTunnel();
```

Comment:
The documentation says there is a setting to control this behavior, but I don't see an if.